### DISCUSSION TITLE
Design a recycling bin system

### DISCUSSION THREAD

**Discussion by EmmanuelMess** [TeamAmaze/AmazeFileManager#2782] [Ideas]
**Describe the bug**
* Where would it be?
* How are files deleted?
* How are files restored?
* How are users differentiated?
* How hidden is the store of files?
* Can the storage of deleted files be accessed from other file managers?
* Does the storage of deleted files have a size limit?
* How is metadata kept?
* Is it enabled by default?

**Comment by EmmanuelMess** [comment]
If the user swaps sd cards the files should stay in the sd card otherwise it might create a weird situation where the files are on the phone, but the sd card is not connected.
To keep track of the user id we can use a ".Trash-<id>" format, like Ubuntu.
To keep track of metadata we can use a separate plain text file, like Ubuntu.

**Comment by VishalNehra** [comment]
@EmmanuelMess Remind me again why we need metadata file.

**Comment by EmmanuelMess** [comment]
> Remind me again why we need metadata file.

Some metadata gets modified on move afaik (like timestamps), and the need to save the restoration location.

**Comment by VishalNehra** [comment]
@EmmanuelMess Right correct - for restoration location. How about actually preserving last modified for move operation?

**Comment by EmmanuelMess** [comment]
> Right correct - for restoration location. How about actually preserving last modified for move operation?

The operation `restore` must revert all changes from the `delete`. As the `delete` operation does a `move` all changes made by the `move` must be undone by `restore`. That includes some timestamps.

**Comment by VishalNehra** [comment]
@EmmanuelMess Yes. That's why I mentioned we should preserve created and modified date with move operation to begin with. I'm not sure what other timestamps you are referring to.
Maintaining more fields other than restore location should make sense, but I'll leave it upto the person who does the POC.

**Comment by EmmanuelMess** [comment]
> Yes. That's why I mentioned we should preserve created and modified date with move operation to begin with. I'm not sure what other timestamps you are referring to.

Ah, I misunderstood then. Anyways, a test for the entire metadata should be written, so that we know nothing failed weirdly.

> Maintaining more fields other than restore location should make sense, but I'll leave it upto the person who does the POC.

That is a good idea

**Comment by VishnuSanal** [comment]
> * Where would it be?
> * How are files deleted?
> * How are files restored?
> * How are users differentiated?
> * How hidden is the store of files?
> * Can the storage of deleted files be accessed from other file managers?
> * Does the storage of deleted files have a size limit?
> * How is metadata kept?
> * Is it enabled by default?

@VishalNehra @EmmanuelMess @TranceLove @shuvashish76 Please give your thoughts on these

**Reply by EmmanuelMess** [reply]
See how Ubuntu does recycling bin.

**Reply by VishnuSanal** [reply]
https://github.com/andreafrancia/trash-cli

**Reply by shuvashish76** [reply]
FOSS app with Recycle Bin feature : https://f-droid.org/en/packages/com.github.axet.filemanager/

**Comment by VishalNehra** [comment]
Library is up : https://github.com/TeamAmaze/AmazeTrashBin
